### PR TITLE
Introduces Taxonomy Flow

### PR DESCRIPTION
**Pull request by lferreira**
This is core implementation from Taxonomy Flow based on Engines.
  You can't access UnitType maps anymore and this implementation is the
   first step to encapsulate the other mpas.

ref: https://github.com/GrupoZapVivaReal/squad-growth/issues/658

### CODE REVIEW

**Comment by brennovich** [comment]
This PR no closes https://github.com/GrupoZapVivaReal/zagallo/pull/27 as https://github.com/GrupoZapVivaReal/zagallo/pull/47 and https://github.com/GrupoZapVivaReal/zagallo/pull/48 are already merged ;)

**Comment by lferreira** [comment]
yeah , is just ref

File: segment_vivareal.go [hunk 1]
```diff
@@ -131,31 +131,31 @@ var vivarealSegmentCog = map[string]uriSemantic{
...
 							levelNames = append(levelNames, uricategory.UriCategory_STATE.String())
 
 							continue
 						}
 					}
 				}
 
-				if _, exists := UnitTypeToEnum[part]; exists {
-					segment.UnitType = part
+				if Taxonomy(segm.Engine).UnitType.Exists(part) {
```
Comment (brennovich) [inline segment_vivareal.go]: Verry cool this `Exists` method, I think.

File: taxonomy.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package zagallo
+
+type i18n struct {
```
Comment (brennovich) [inline taxonomy.go]: I would add documentation here to guide further contributors (including ourselves) about motivations, what this really means, and warning about including things like previous `Weight` field.

Comment (brennovich) [reply taxonomy.go]: Here are some docs that I put in my hackweek PR: https://github.com/GrupoZapVivaReal/zagallo/pull/27/commits/732fb7580db98ac1f3dc8df6fd5824533d43bafc#diff-e33944ecab5e2a0cdd35e0bc008bb0f0 could be useful **in this file**

File: taxonomy_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package zagallo
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/vivareal/proto-resources/listings/enums"
+)
+
+func TestTaxonomy(t *testing.T) {
```
Comment (brennovich) [inline taxonomy_test.go]: Such test https://github.com/GrupoZapVivaReal/zagallo/pull/27/commits/732fb7580db98ac1f3dc8df6fd5824533d43bafc#diff-867ca3065676d338a89833104ccebf7bR9 would ensure that things are properly defined (like https://github.com/GrupoZapVivaReal/zagallo/blob/master/unit_type_test.go does but relying on engines). I think we could also include this tests ;)

File: unit_type.go [hunk 4]
```diff
@@ -16,248 +16,32 @@ type Dictionary struct {
...
+	}
+	return false
+}
+
+// Returns I18n context from a listingsEnums.UnitType
+func (utt unitTypeTaxonomy) I18n(unit listingsEnums.UnitType) i18n {
+	return utt.i18nMap[unit]
+}
+
+// Returns listingsEnums.UnitType related to a uri part
```
Comment (brennovich) [inline unit_type.go]: Golang has this stupid convention of include name of the `function` in the doc itself: https://blog.golang.org/godoc-documenting-go-code can we follow it? I'm trying to adopt it in our projects, `zagallo` is the one that has most messed up docs, `romario` new's function are much better documented https://github.com/GrupoZapVivaReal/romario/blob/master/pkg/widget/widget.go#L196

File: unit_type_vivareal.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
+package zagallo
+
+import (
+	"github.com/vivareal/proto-resources/listings/enums"
+)
+
+var uriUnitTypeVivaRealToEnum = map[string]listingsEnums.UnitType{
```
Comment (brennovich) [inline unit_type_vivareal.go]: What do you think about using this convention of prefixing portal and sufixing `cog`? https://github.com/GrupoZapVivaReal/zagallo/blob/master/segment_vivareal.go#L12

File: listingpage.go [hunk 6]
```diff
@@ -126,8 +126,11 @@ func buildTitle(resource *listings.ListingResource) string {
 	area := calculateArea(resource)
 	bedrooms := int32biggestOne(resource.Bedrooms)
 
+	//TODO: Implements Engine by Portal
+	txn := Taxonomy(Engines.VivaReal)
```
[the remark this answers is not in the document]
Comment (lferreira) [reply listingpage.go]: we should remember to remove the hardcoded Engine when we'll implement listingpage for ZAP

File: fields_test.go [hunk 7]
```diff
@@ -138,22 +138,22 @@ var pageFieldsValueScenarios = []struct {
 		scenarios: []fieldValueScenario{
 			{
 				about:    "correct value for UF",
-				input:    segment{State: "RN"},
+				input:    segment{Engine: Engines.VivaReal, State: "RN"},
```
[the remark this answers is not in the document]
Comment (lferreira) [reply fields_test.go]: I think I'll do this now

[the remark this answers is not in the document]
Comment (brennovich) [reply fields_test.go]: :1st_place_medal:

File: listingpage.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (brennovich) [reply listingpage.go]: Cool. When we add proper engine support here most of this file will go to `_vivareal`

**Pull request merged**